File: Test Case Generation/deepseek/cpp/test_task_8.cpp
#include <gtest/gtest.h>
#include <vector>
#include "sum_product.h"  // Assume the function is declared here

using namespace std;

// Test fixture for sum_product function
class SumProductTest : public ::testing::Test {
protected:
    void SetUp() override {}
    void TearDown() override {}
};

// Test empty vector case
TEST_F(SumProductTest, EmptyVector) {
    vector<int> numbers = {};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 0);  // Sum of empty vector should be 0
    EXPECT_EQ(result[1], 1);  // Product of empty vector should be 1
}

// Test single element vector
TEST_F(SumProductTest, SingleElement) {
    vector<int> numbers = {5};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 5);
    EXPECT_EQ(result[1], 5);
}

// Test multiple positive integers
TEST_F(SumProductTest, MultiplePositiveIntegers) {
    vector<int> numbers = {1, 2, 3, 4};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 10);  // 1+2+3+4 = 10
    EXPECT_EQ(result[1], 24);  // 1*2*3*4 = 24
}

// Test with zero (product edge case)
TEST_F(SumProductTest, WithZero) {
    vector<int> numbers = {5, 0, 3};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 8);   // 5+0+3 = 8
    EXPECT_EQ(result[1], 0);   // 5*0*3 = 0
}

// Test with negative integers
TEST_F(SumProductTest, WithNegativeIntegers) {
    vector<int> numbers = {-2, -3, -4};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], -9);  // (-2)+(-3)+(-4) = -9
    EXPECT_EQ(result[1], -24); // (-2)*(-3)*(-4) = -24
}

// Test with mixed positive and negative integers
TEST_F(SumProductTest, MixedPositiveNegative) {
    vector<int> numbers = {-1, 2, -3, 4};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 2);   // (-1)+2+(-3)+4 = 2
    EXPECT_EQ(result[1], 24);  // (-1)*2*(-3)*4 = 24
}

// Test large values (potential overflow is implementation-defined)
TEST_F(SumProductTest, LargeNumbers) {
    vector<int> numbers = {1000000, 2000000, 3000000};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 6000000);  // 1e6+2e6+3e6 = 6e6
    EXPECT_EQ(result[1], 6000000000000LL);  // Product may overflow int on some systems
}

// Test with single zero element
TEST_F(SumProductTest, SingleZero) {
    vector<int> numbers = {0};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 0);
    EXPECT_EQ(result[1], 0);
}

// Test with alternating signs
TEST_F(SumProductTest, AlternatingSigns) {
    vector<int> numbers = {1, -1, 2, -2, 3, -3};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 0);   // 1-1+2-2+3-3 = 0
    EXPECT_EQ(result[1], -36); // 1*(-1)*2*(-2)*3*(-3) = -36
}

// Test with all same positive numbers
TEST_F(SumProductTest, AllSamePositive) {
    vector<int> numbers = {2, 2, 2, 2};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 8);   // 2+2+2+2 = 8
    EXPECT_EQ(result[1], 16);  // 2*2*2*2 = 16
}

// Test with all same negative numbers (even count)
TEST_F(SumProductTest, AllSameNegativeEven) {
    vector<int> numbers = {-3, -3, -3, -3};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], -12); // (-3)+(-3)+(-3)+(-3) = -12
    EXPECT_EQ(result[1], 81);  // (-3)*(-3)*(-3)*(-3) = 81
}

// Test with all same negative numbers (odd count)
TEST_F(SumProductTest, AllSameNegativeOdd) {
    vector<int> numbers = {-3, -3, -3};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], -9);  // (-3)+(-3)+(-3) = -9
    EXPECT_EQ(result[1], -27); // (-3)*(-3)*(-3) = -27
}

// Test with maximum number of elements (performance test)
TEST_F(SumProductTest, ManyElements) {
    vector<int> numbers(1000, 1);  // 1000 elements of 1
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 1000);  // Sum of 1000 ones = 1000
    EXPECT_EQ(result[1], 1);     // Product of 1000 ones = 1
}

// Test with two elements
TEST_F(SumProductTest, TwoElements) {
    vector<int> numbers = {5, 7};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 12);
    EXPECT_EQ(result[1], 35);
}

// Test with increasing sequence
TEST_F(SumProductTest, IncreasingSequence) {
    vector<int> numbers = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 55);   // Arithmetic series sum
    EXPECT_EQ(result[1], 3628800);  // 10!
}

// Test with descending sequence
TEST_F(SumProductTest, DescendingSequence) {
    vector<int> numbers = {10, 9, 8, 7, 6, 5, 4, 3, 2, 1};
    vector<int> result = sum_product(numbers);
    
    EXPECT_EQ(result.size(), 2);
    EXPECT_EQ(result[0], 55);
    EXPECT_EQ(result[1], 3628800);
}

File: Test Case Generation/deepseek/python/test_task_8.py
import pytest
from typing import List, Tuple
from src.sum_product import sum_product  # Adjust import path as needed

class TestSumProduct:
    """Test suite for sum_product function."""
    
    def test_empty_list(self):
        """Test that empty list returns (0, 1)."""
        assert sum_product([]) == (0, 1)
    
    def test_single_element(self):
        """Test list with single element."""
        assert sum_product([5]) == (5, 5)
        assert sum_product([-3]) == (-3, -3)
        assert sum_product([0]) == (0, 0)
    
    def test_two_elements(self):
        """Test list with two elements."""
        assert sum_product([2, 3]) == (5, 6)
        assert sum_product([-1, 5]) == (4, -5)
        assert sum_product([0, 10]) == (10, 0)
    
    def test_multiple_elements(self):
        """Test list with multiple elements."""
        assert sum_product([1, 2, 3, 4]) == (10, 24)
        assert sum_product([1, 1, 1, 1]) == (4, 1)
        assert sum_product([-1, -2, -3, -4]) == (-10, 24)
    
    def test_with_zeros(self):
        """Test lists containing zeros."""
        assert sum_product([0, 1, 2, 3]) == (6, 0)
        assert sum_product([0, 0, 5]) == (5, 0)
        assert sum_product([0, 0, 0]) == (0, 0)
    
    def test_with_negative_numbers(self):
        """Test lists with negative numbers."""
        assert sum_product([-1, -2, 3]) == (0, 6)
        assert sum_product([-5, -5, -5]) == (-15, -125)
        assert sum_product([-1, 2, -3, 4]) == (2, 24)
    
    def test_large_numbers(self):
        """Test with large integers."""
        assert sum_product([1000, 2000, 3000]) == (6000, 6000000000)
        assert sum_product([-1000, 1000]) == (0, -1000000)
    
    def test_product_overflow_behavior(self):
        """Test with numbers that might cause overflow (Python handles big ints)."""
        result = sum_product([2, 3, 5, 7, 11, 13, 17, 19])
        expected_sum = 2 + 3 + 5 + 7 + 11 + 13 + 17 + 19
        expected_product = 2 * 3 * 5 * 7 * 11 * 13 * 17 * 19
        assert result == (expected_sum, expected_product)
    
    def test_single_zero(self):
        """Test single zero element."""
        assert sum_product([0]) == (0, 0)
    
    def test_single_negative_zero(self):
        """Test single negative zero (identical to zero)."""
        assert sum_product([-0]) == (0, 0)
    
    def test_alternating_signs(self):
        """Test alternating positive and negative numbers."""
        assert sum_product([-1, 2, -3, 4, -5]) == (-3, -120)
        assert sum_product([1, -2, 3, -4, 5]) == (3, -120)
    
    def test_very_large_list(self):
        """Test with a large list to verify performance and correctness."""
        numbers = list(range(1, 101))  # 1 to 100
        expected_sum = 5050  # Sum of 1..100
        # Product is huge, but Python handles it
        result = sum_product(numbers)
        assert result[0] == expected_sum
        # Only verify product has correct sign (positive since all numbers positive)
        assert result[1] > 0
    
    def test_repeated_same_number(self):
        """Test list with repeated same number."""
        assert sum_product([3, 3, 3]) == (9, 27)
        assert sum_product([-2, -2, -2, -2]) == (-8, 16)
    
    def test_one_and_negative_one_combinations(self):
        """Test combinations of 1 and -1."""
        assert sum_product([1, -1, 1, -1]) == (0, 1)
        assert sum_product([-1, -1, -1]) == (-3, -1)
    
    def test_mixed_positive_negative_with_zero(self):
        """Test mixed signs including zero."""
        assert sum_product([-1, 0, 1, -2, 2]) == (0, 0)
        assert sum_product([-5, 3, 0, -2, 4]) == (0, 0)